File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
import java.lang.*;
class Solution {
    private static class State {
    public final int index;
    public final int acc;
    State(int index, int acc) {
        this.index = index;
        this.acc = acc;
    }
}
    public int solution(int[] numbers, int target) {
        int count = 0;
        Stack<State> s = new Stack<>();
        s.push(new State(0,0));
        while(!s.isEmpty()){
            State state = s.pop();
            if(state.index == numbers.length) {
                if(state.acc == target) count++;
                continue;
            }
            s.push(new State(state.index + 1, state.acc - numbers[state.index]));
            s.push(new State(state.index + 1, state.acc + numbers[state.index]));
        }
        return count;
    }
}

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.js
function solution(numbers, target) {
    // 초기값 0,0 으로 stack 만들기
    const stack = [new State(0,0)];
    let count = 0;
    while(stack.length > 0) {
        const state = stack.pop();
        if(state.index === numbers.length) {
            if(state.acc === target) count++;
            continue;
        }
        // +일경우
        stack.push(new State(state.index + 1, state.acc - numbers[state.index]));
        // -일경우
        stack.push(new State(state.index + 1, state.acc + numbers[state.index]));
    }
    return count;
}

class State {
    constructor(index, acc) {
        this.index = index;
        this.acc = acc;
    }
}